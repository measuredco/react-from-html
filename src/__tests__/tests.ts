/* eslint-env jest */
import * as React from "react";
import reactFromHtml from "..";

describe("reactFromHtml E2E tests", async () => {
  it("Should hydrate a basic component", async () => {
    const componentName: string = "myComponent";

    const hydrator = async () => {
      return React.createElement("span", {}, "hydrated component");
    };

    const hydrators = { [`.${componentName}`]: hydrator };
    const documentElement = document.createElement("div");

    documentElement.innerHTML = `<div class="${componentName}"></div>`;

    await reactFromHtml(documentElement, hydrators);

    expect(documentElement.innerHTML).toMatchSnapshot();
  });

  it("Should work for nested hydratables", async () => {
    const mockCall = jest.fn();
    const hydrators = {
      [`.Parent`]: async (el, hydrate) => {
        mockCall();

        return React.createElement("span", {}, await hydrate(el));
      },
      [`.Child`]: async (el, hydrate) => {
        mockCall();

        return React.createElement("b", {}, await hydrate(el));
      },
    };

    const documentElement = document.createElement("div");

    documentElement.innerHTML = `
      <div class="Parent">
        <div class="Child">
          Hello, World!
        </div>
      </div>
      `;

    await reactFromHtml(documentElement, hydrators);

    expect(documentElement.innerHTML).toMatchSnapshot();
    expect(mockCall).toBeCalledTimes(2);
  });

  it("Should hydrate components with custom query selectors", async () => {
    const componentName: string = "myComponent";

    const hydrator = async () => {
      return React.createElement("span", {}, "hydrated component");
    };

    const hydrators = { [componentName]: hydrator };
    const documentElement = document.createElement("div");

    documentElement.innerHTML = `<div data-hydratable="test-${componentName}"></div>`;

    await reactFromHtml(documentElement, hydrators, {
      extra: {},
      getQuerySelector: key => `[data-hydratable*="${key}"]`,
    });

    expect(documentElement.innerHTML).toMatchSnapshot();
  });
});
