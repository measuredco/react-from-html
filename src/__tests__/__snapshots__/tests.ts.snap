// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reactFromHtml E2E tests Should hydrate a basic component 1`] = `"<div class=\\"hydration-root\\" data-hydration-id=\\"1\\"><span>hydrated component</span></div>"`;

exports[`reactFromHtml E2E tests Should hydrate components with custom query selectors 1`] = `"<div class=\\"hydration-root\\" data-hydration-id=\\"4\\"><span>hydrated component</span></div>"`;

exports[`reactFromHtml E2E tests Should work for nested hydratables 1`] = `
"
      <div class=\\"hydration-root\\" data-hydration-id=\\"2\\"><span>
        <b>
          Hello, World!
        </b>
      </span></div>
      "
`;
